### PR TITLE
Rewrite dependency parsing and load order

### PR DESCRIPTION
**Pull request by RustyBlade64**
This PR rewrites mod dependency parsing and mod load order. Both hidden optional (`(?)`) as well as conflicting (`!`) dependencies are handled now. Additionally mod load order has been changed to match in-game behaviour in what hopefully is nearly all cases (see breakdown below).

The only thing that should be missing now is to use natural sort order instead of alphabetic one. However, there are several conflicting definitions for natural sort order and there is no documentation which one Factorio is using. It should only make a difference in very rare corner cases anyway. So I decided to leave this out.

This PR fixes points two and three of #10 and is a superset of #11. It also fixes the first of multiple issues from KirkMcDonald/factorio-tools#12.

## Before/after load order comparison

TLDR In both cases new load order matches in-game load order.

#### Dependency tree
* base
  * fe-core
    * fe-equipment
    * fe-logistics
    * fe-power
    * fe-transport
      * fe-machines
      * fe-storage
    * fe-weaponry

#### Game load order
```
Checksum for core: 2087614386
Checksum of base: 1061071205
Checksum of FactorioExtended-Plus-Core: 810350191
Checksum of FactorioExtended-Plus-Equipment: 2917099853
Checksum of FactorioExtended-Plus-Logistics: 3782117776
Checksum of FactorioExtended-Plus-Power: 18660506
Checksum of FactorioExtended-Plus-Transport: 1338988286
Checksum of FactorioExtended-Plus-Weaponry: 1056291079
Checksum of FactorioExtended-Plus-Machines: 1568488487
Checksum of FactorioExtended-Plus-Storage: 749124387
```

#### Original load order
```lua
{
  "core",
  "base",
  "FactorioExtended-Plus-Core",
  "FactorioExtended-Plus-Equipment",
  "FactorioExtended-Plus-Logistics",
  "FactorioExtended-Plus-Machines", -- should load after transport
  "FactorioExtended-Plus-Power",
  "FactorioExtended-Plus-Transport",
  "FactorioExtended-Plus-Storage",
  "FactorioExtended-Plus-Weaponry"
} --[[table: 0x55f200fc7fd0]]
```
_Machines_ loading before _Transport_ was a bug in the old implementation where the dependency string was parsed incorrectly.

#### New load order
```lua
{
  "core",
  "base",
  "FactorioExtended-Plus-Core",
  "FactorioExtended-Plus-Equipment",
  "FactorioExtended-Plus-Logistics",
  "FactorioExtended-Plus-Power",
  "FactorioExtended-Plus-Transport",
  "FactorioExtended-Plus-Weaponry",
  "FactorioExtended-Plus-Machines",
  "FactorioExtended-Plus-Storage"
} --[[table: 0x5595f3b3ba20]]
```

#### Dependency DAG
seablock-dependency-dag
This is the load order I get when going through the `info.json` files by hand. For it to be more readable I have stripped out all explicit transitive dependencies as well as the dependencies of _SeaBlockMetaPack_ (it simply depends on every single mod in the mod-pack).

The green edges do not have any special meaning. I chose a separate color to make it clearer where each of them goes in cases of overlapping lines.

#### Game load order
```
Checksum for core: 2087614386
Checksum of base: 1061071205
Checksum of A Sea Block Config: 4016453794
Checksum of bobenemies: 1060662388
Checksum of bobinserters: 1186143822
Checksum of boblibrary: 201512343
Checksum of Explosive Excavation: 2284750772
Checksum of FNEI: 947672031
Checksum of KS_Power: 2533975514
Checksum of Nuclear Fuel: 1817035091
Checksum of ScienceCostTweakerM: 2035503171
Checksum of bobequipment: 1747497181
Checksum of bobplates: 1807866958
Checksum of bobtech: 834284230
Checksum of bobassembly: 3467163679
Checksum of bobelectronics: 4078712591
Checksum of boblogistics: 3500215993
Checksum of bobmining: 1985073245
Checksum of bobpower: 2431287710
Checksum of bobrevamp: 988316657
Checksum of bobwarfare: 1946742629
Checksum of angelsrefining: 1087409800
Checksum of bobmodules: 2483356921
Checksum of angelsaddons-warehouses: 2217531831
Checksum of angelspetrochem: 259098712
Checksum of CircuitProcessing: 4003027249
Checksum of LandfillPainting: 2995760538
Checksum of SpaceMod: 4257687145
Checksum of angelssmelting: 609360190
Checksum of angelsbioprocessing: 3387071644
Checksum of SeaBlock: 3320564922
Checksum of SeaBlockMetaPack: 0
```

#### Original load order
```lua
{
  "core",
  "base",
  "A Sea Block Config",
  "boblibrary",
  "bobenemies",
  "bobplates",
  "bobelectronics",
  "bobmodules",
  "bobrevamp",
  "CircuitProcessing",
  "Explosive Excavation",
  "FNEI",
  "KS_Power",
  "angelsrefining",
  "LandfillPainting",
  "Nuclear Fuel",
  "ScienceCostTweakerM",
  "bobassembly",
  "angelspetrochem",
  "angelssmelting",
  "angelsbioprocessing",
  "bobinserters",
  "boblogistics",
  "bobtech",
  "bobequipment",
  "SpaceMod",
  "SeaBlock",
  "angelsaddons-warehouses",
  "bobmining",
  "bobpower",
  "bobwarfare",
  "SeaBlockMetaPack"
} --[[table: 0x55aeaa9eb710]]
```
Albeit correct, this load order is completely different from in-game behaviour.

#### New load order
```lua
{
  "core",
  "base",
  "A Sea Block Config",
  "bobenemies",
  "bobinserters",
  "boblibrary",
  "Explosive Excavation",
  "FNEI",
  "KS_Power",
  "Nuclear Fuel",
  "ScienceCostTweakerM",
  "bobequipment",
  "bobplates",
  "bobtech",
  "bobassembly",
  "bobelectronics",
  "boblogistics",
  "bobmining",
  "bobpower",
  "bobrevamp",
  "bobwarfare",
  "angelsrefining",
  "bobmodules",
  "angelsaddons-warehouses",
  "angelspetrochem",
  "CircuitProcessing",
  "LandfillPainting",
  "SpaceMod",
  "angelssmelting",
  "angelsbioprocessing",
  "SeaBlock",
  "SeaBlockMetaPack"
} --[[table: 0x55d7cb3de0b0]]
```

### CODE REVIEW

**Comment by KirkMcDonald** [comment]
Apologies for neglecting this project for so long, but thank you very much for this PR.
